### PR TITLE
:new: Add GrandQC Tissue Segmentation Model

### PR DESCRIPTION
**Pull request by Jiaqi-Lv**
## 🚀Summary
This PR introduces a new **[GrandQC Tissue Detection Model](https://github.com/cpath-ukk/grandqc/tree/main)** for digital pathology quality control and integrates **EfficientNet-based encoder architecture** into the TIAToolbox framework.

---

## ✨Key Changes
- **New Model Architecture**
  - Added `grandqc.py` implementing a UNet++ decoder with EfficientNet encoder for tissue segmentation.
  - Includes preprocessing (JPEG compression + ImageNet normalization), postprocessing (argmin-based mask generation), and batch inference utilities.
- **EfficientNet Encoder**
  - Added `timm_efficientnet.py` providing configurable EfficientNet encoders with dilation support and custom input channels.
- **Pretrained Model Config**
  - Updated `pretrained_model.yaml` to register `grandqc_tissue_detection_mpp10` with associated IO configuration.
  - Corrected `IOSegmentorConfig` references and adjusted resolutions for SCCNN models.
- **Testing**
  - Added comprehensive unit tests for:
    - `GrandQCModel` functionality, preprocessing/postprocessing, and decoder blocks.
    - EfficientNet encoder utilities and scaling logic.
  
## Impact
- Enables high-resolution tissue detection for WSI quality control using state-of-the-art architectures.
- Improves flexibility for segmentation tasks with EfficientNet encoders.
- Enhances code quality and consistency through updated linting and formatting tools.

## Tasks
- [x] Re-host GrandQC model weights on TIA Hugging Face
- [x] Update `pretrained_model.yaml`
- [x] Update `requirements.txt`
- [x] Define GrandQC model architecture
- [x] Add example usage
- [x] Remove segmentation-models-pytorch dependency
- [x] Wait for response from GrandQC authors
- [x] Add tests
- [x] Tidy up

### CODE REVIEW

File: tiatoolbox/models/architecture/grandqc.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+        >>> results = semantic_segmentor.run(
+        ...     ["/example_wsi.svs"],
+        ...     masks=None,
+        ...     auto_get_mask=False,
+        ...     patch_mode=False,
+        ...     save_dir=Path("/tissue_mask/"),
+        ...     output_type="annotationstore",
+        ... )
+
+    """
```
Comment (shaneahmed) [inline tiatoolbox/models/architecture/grandqc.py:540]: Please add citation to GrandQC in the documentation.

File: tiatoolbox/models/architecture/grandqc.py [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+import cv2
+import numpy as np
+import segmentation_models_pytorch as smp
+import torch
+
+from tiatoolbox.models.models_abc import ModelABC
+
+
+class TissueDetectionModel(ModelABC):
```
Comment (shaneahmed) [inline tiatoolbox/models/architecture/grandqc.py]:
```suggestion
class GrandQCModel(ModelABC):
```

**Comment by shaneahmed** [comment]
GrandQC authors have confirmed that we are allowed to integrate the model in TIAToolbox. Please add appropriate licensing and citation information.

File: tests/models/test_arch_grandqc.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+    GrandQCModel,
+    SegmentationHead,
+    UnetPlusPlusDecoder,
+)
+from tiatoolbox.models.engine.io_config import IOSegmentorConfig
+from tiatoolbox.utils.misc import select_device
+from tiatoolbox.wsicore.wsireader import VirtualWSIReader
+
+ON_GPU = False
+
```
Comment (shaneahmed) [inline tests/models/test_arch_grandqc.py:28]: Add a test with sample_svs as input. Count the number of pixels in the foreground in the output. Check the sum of foreground pixels in the output.

File: tiatoolbox/data/pretrained_model.yaml [hunk 4]
```diff
@@ -934,3 +934,21 @@ nuclick_light-pannuke:
             patch_input_shape: [128, 128]
             patch_output_shape: [128, 128]
             save_resolution: {'units': 'baseline', 'resolution': 1.0}
+
+grandqc_tissue_detection_mpp10:
```
Comment (shaneahmed) [inline tiatoolbox/data/pretrained_model.yaml]:
```suggestion
grandqc_tissue_detection:
```

**Review by shaneahmed — COMMENTED** [review]
Thanks @Jiaqi-Lv 
I have updated docstrings in this commit. 
https://github.com/TissueImageAnalytics/tiatoolbox/pull/965/commits/1895e38c3ccfbd0564bd5ad9c1e2d8fdac1ce44a

Can you check these?

File: tiatoolbox/models/architecture/grandqc.py [hunk 5]
```diff
@@ -0,0 +1,691 @@
...
+            out_channels,
+            out_channels,
+            kernel_size=3,
+            padding=1,
+        )
+        self.attention2 = nn.Identity()
+
+    def forward(
+        self: DecoderBlock,
+        x: torch.Tensor,
```
Comment (shaneahmed) [inline tiatoolbox/models/architecture/grandqc.py]:
```suggestion
        input_tensor: torch.Tensor,
```

File: tiatoolbox/models/architecture/grandqc.py [hunk 6]
```diff
@@ -0,0 +1,691 @@
...
+        skip: torch.Tensor | None = None,
+    ) -> torch.Tensor:
+        """Forward pass through the decoder block.
+
+        Upsamples the input tensor, concatenates it with the skip connection
+        (if provided), and applies two convolutional layers with attention.
+
+        Args:
+            x (torch.Tensor):
+                Input tensor from the previous decoder layer.
```
Comment (shaneahmed) [inline tiatoolbox/models/architecture/grandqc.py]: Please can you also specify expected shape?

File: tiatoolbox/models/architecture/timm_efficientnet.py [hunk 7]
```diff
@@ -0,0 +1,651 @@
...
+        }
+
+    def forward(self, x: torch.Tensor) -> list[torch.Tensor]:
+        """Forward pass through EfficientNet encoder.
+
+        Extracts feature maps from multiple stages of the encoder for use in
+        decoder networks or multi-scale processing.
+
+        Args:
+            x (torch.Tensor):
```
Comment (shaneahmed) [inline tiatoolbox/models/architecture/timm_efficientnet.py:481]: Please rename x to something meaningful.

File: tiatoolbox/models/architecture/grandqc.py [hunk 8]
```diff
@@ -0,0 +1,698 @@
...
+        conv2 = Conv2dReLU(
+            out_channels,
+            out_channels,
+            kernel_size=3,
+            padding=1,
+        )
+        super().__init__(conv1, conv2)
+
+
+class UnetPlusPlusDecoder(nn.Module):
```
Comment (shaneahmed) [inline tiatoolbox/models/architecture/grandqc.py:361]: I am not sure why this is part of grandqc.

**Review by adamshephard — COMMENTED** [review]
Looks great. Can we also add in an argument to the WSIReader get tissue mask to use grandqc automatically?

File: tiatoolbox/models/architecture/grandqc.py [hunk 9]
```diff
@@ -0,0 +1,698 @@
+"""GrandQC Tissue Detection Model Architecture.
+
```
Comment (adamshephard) [inline tiatoolbox/models/architecture/grandqc.py:2]: We will need to add reference to github repo and paper here.

**Comment by adamshephard** [comment]
I think we should also add in references (to the github repo and paper and license) in both the code and the documentation. Encouraging people to cite the paper if they use it.

File: tiatoolbox/models/architecture/grandqc.py [hunk 10]
```diff
@@ -0,0 +1,698 @@
...
+This module defines the GrandQC model for tissue detection in digital pathology.
+It implements a UNet++ architecture with an EfficientNetB0 encoder and a segmentation
+head for high-resolution tissue segmentation. The model is designed to identify
+tissue regions and background areas for quality control in whole slide images (WSIs).
+
```
Comment (shaneahmed) [inline tiatoolbox/models/architecture/grandqc.py:8]: Please cite the paper [1], if you use this model.

File: tiatoolbox/models/architecture/grandqc.py [hunk 11]
```diff
@@ -0,0 +1,698 @@
...
+    ...     ["/example_wsi.svs"],
+    ...     masks=None,
+    ...     auto_get_mask=False,
+    ...     patch_mode=False,
+    ...     save_dir=Path("/tissue_mask/"),
+    ...     output_type="annotationstore",
+    ... )
+
+References:
+    Weng, Zhilong et al. "GrandQC: A comprehensive solution to quality control
```
Comment (shaneahmed) [inline tiatoolbox/models/architecture/grandqc.py]:
```suggestion
    [1] Weng, Zhilong et al. "GrandQC: A comprehensive solution to quality control
```

File: tiatoolbox/models/architecture/grandqc.py [hunk 12]
```diff
@@ -0,0 +1,699 @@
...
+
+Features:
+---------
+- JPEG compression and ImageNet normalization during preprocessing.
+- Argmin-based postprocessing for generating tissue masks.
+- Efficient inference pipeline for batch processing.
+
+Example:
+    >>> from tiatoolbox.models.engine.semantic_segmentor import SemanticSegmentor
+    >>> segmentor = SemanticSegmentor(model="grandqc_tissue_detection_mpp10")
```
Comment (measty) [inline tiatoolbox/models/architecture/grandqc.py:32]: need to change example to:
segmentor = SemanticSegmentor(model="grandqc_tissue_detection")
to match model name in .yaml

File: tiatoolbox/models/architecture/grandqc.py [hunk 13]
```diff
@@ -0,0 +1,699 @@
...
+                Size of the convolution kernel.
+            padding (int):
+                Padding applied to the input. Defaults to 0.
+            stride (int):
+                Stride of the convolution. Defaults to 1.
+
+        """
+        norm = nn.BatchNorm2d(out_channels)
+
+        is_identity = isinstance(norm, nn.Identity)
```
Comment (measty) [inline tiatoolbox/models/architecture/grandqc.py]: given norm is defined above as a nn.BatchNorm2d instance, why is this check needed? it will always be false right?

**Review by measty — APPROVED** [review]
Looks good, just a few small comments

File: tiatoolbox/models/architecture/grandqc.py (hunk 12 above)
Comment (shaneahmed) [reply tiatoolbox/models/architecture/grandqc.py:32]: Fixed in https://github.com/TissueImageAnalytics/tiatoolbox/pull/965/commits/1ab6728a69d423dc06750f721175c92a37f619ee

File: tiatoolbox/models/architecture/grandqc.py (hunk 13 above)
Comment (shaneahmed) [reply tiatoolbox/models/architecture/grandqc.py]: Fixed in https://github.com/TissueImageAnalytics/tiatoolbox/pull/965/commits/1ab6728a69d423dc06750f721175c92a37f619ee and https://github.com/TissueImageAnalytics/tiatoolbox/pull/965/commits/cc4499a6e86bbf099aecc82c5f484023ae22f629
